proposals/plugins.md

```diff
@@ -0,0 +1,146 @@
+   * Some of this will likely be provided by CloudFormation but kube-aws could possibly replicas the checks to save time for the stack create/update to fail
+
+# Design
+
+## Plugin Types
+
+* Built-in plugins versus ad-hoc user specific plugins
+  * Some plugins will be provided by kube-aws and some will be specific to a user
+  * Both will use the same mechanisms to hook into kube-aws/deploy etc
+  * Both types of plugins will be enabled and configured in the same way
+* Built-in plugins versus contrib plugins
+  * kube-aws will accept contrib plugins for features considered to be applicable to many users
+  * Contrib plugins will be enabled in a similar way but specifically marked as not fully
+  battle tested; they will still be configured in the same way
+* Plugins will be allowed to act across the whole cluster or act on a specific set of node pools, this should form part of the base configuration for which plugins to run
+
+## Stages
+
+The workflow for kube-aws should consist of the following stages, each of which will be called on the plugins in turn. Each stage will complete across all plugins prior to proceeding to the next stage. All plugins will receive input based on the whole setup, not just their own. This will allow them to hook into parameters from other plugins or core logic. The stages are not intended to become kube-aws commands and should mostly be transparent to a kube-aws user.
+
+1. initialise
+   * Perform any prerequisite setup required such as calls to external system. Configuration, presence of files etc can be validated.
+   * Input - configuration
+   * Output - configuration with any initialised values, internal kube-aws model with primary constructs for CloudFormation, Kubernetes YAML/JSON, Helm charts, systemd units, kube API server parameters and any other key areas that allow customisation. Some of these will be file references but some will be full kube-aws internal models. The available list of constructs will be added to in further implementations.
+1. render
+   * Stack files are gathered and rendered locally, most plugins will perform a no-op here but some may choose to output additional CloudFormation stacks or perform additional validation. A core plugin could grab all the primary constructs and create the main CloudFormation or a kube API server plugin could validate none of the parameters it has set are conflicting.
+   * Input - output of `initialise` stage
+   * Output - CloudFormation file(s), other file reference(s) such as Helm Charts
+1. deploy
+   * Take the CloudFormation and file(s), upload the files to S3 and deploy the template(s) to CloudFormation. This is an opportunity for plugins to perform waiting tasks and/or actions that need to be done on deploy/install, for example post to an endpoint created during deploy.
+   * Input - output of `render` stage, stack name
+   * Output - stack details such as stack name, endpoints, certs
+1. validate
```

Comment:
Happy to rename. I agree we should avoid name clashes in cases we expose this to the users, in this case it will be exposed directly in our operator documentation so worth picking something more descriptive. How about `postDeployValidate`? Or simply `postDeploy` is an option as well if we wish to allow plugins to perform other "post" actions.